File: examples_nodejs/example_keygen.js
/*
 * libzupt - Key Generation Example (Node.js)
 * Shows how to generate, save, load, and manage key pairs
 *
 * SPDX-License-Identifier: MIT
 */

const zupt = require('./index');
const fs = require('fs');
const path = require('path');

function main() {
    console.log('libzupt - Key Generation Example (Node.js)');
    console.log('Library version:', zupt.getVersion());
    console.log('');

    const testDir = '/tmp/libzupt_keygen_test';
    fs.mkdirSync(testDir, { recursive: true });

    try {
        const privKeyFile = path.join(testDir, 'keys.pem');
        const pubKeyFile = path.join(testDir, 'keys.pub.pem');

        // Step 1: Generate new key pair
        console.log('Step 1: Generating new key pair...');
        const keygen = new zupt.KeyGenerator();
        const keypair = keygen.generateKeyPair();

        console.log('  Public key size:', keypair.publicKey.length, 'bytes');
        console.log('  Secret key size:', keypair.secretKey.length, 'bytes');
        console.log('  Public key (hex, first 32 bytes):');
        console.log('  ', keypair.publicKey.slice(0, 32).toString('hex'));

        // Step 2: Save key pair to file
        console.log('');
        console.log('Step 2: Saving key pair to file...');
        keygen.saveKeyPair(keypair, privKeyFile);
        console.log('  Saved to:', privKeyFile);

        // Step 3: Export public key
        console.log('');
        console.log('Step 3: Exporting public key...');
        keygen.exportPublicKey(privKeyFile, pubKeyFile);
        console.log('  Saved to:', pubKeyFile);

        // Step 4: Load public key from file
        console.log('');
        console.log('Step 4: Loading public key from file...');
        const publicKey = keygen.loadPublicKey(pubKeyFile);
        console.log('  Loaded public key size:', publicKey.length, 'bytes');

        // Step 5: Load key pair from file
        console.log('');
        console.log('Step 5: Loading key pair from file...');
        const loadedKeyPair = keygen.loadKeyPair(privKeyFile);
        console.log('  Loaded public key size:', loadedKeyPair.publicKey.length, 'bytes');
        console.log('  Loaded secret key size:', loadedKeyPair.secretKey.length, 'bytes');

        // Step 6: Verify keys match
        console.log('');
        console.log('Step 6: Verifying keys...');
        let keysMatch = true;
        if (loadedKeyPair.publicKey.length !== keypair.publicKey.length) {
            keysMatch = false;
        } else {
            for (let i = 0; i < keypair.publicKey.length; i++) {
                if (loadedKeyPair.publicKey[i] !== keypair.publicKey[i]) {
                    keysMatch = false;
                    break;
                }
            }
        }

        if (keysMatch) {
            console.log('  SUCCESS: Keys match!');
        } else {
            console.log('  ERROR: Keys do not match!');
        }

        // Step 7: Demonstrate key size constants
        console.log('');
        console.log('Step 7: Key sizes:');
        console.log('  ML-KEM-768 Public Key: 1184 bytes');
        console.log('  ML-KEM-768 Secret Key: 2400 bytes');
        console.log('  X25519 Key: 32 bytes');
        console.log('  Hybrid Public Key: 1224 bytes (8 header + 1184 ML + 32 X25519)');
        console.log('  Hybrid Secret Key: ~3664 bytes (includes public key + private keys + checksum)');

        // Step 8: Generate multiple key pairs (demonstrate randomness)
        console.log('');
        console.log('Step 8: Generating 3 key pairs to verify randomness...');
        for (let i = 0; i < 3; i++) {
            const kp = keygen.generateKeyPair();
            const firstBytes = kp.publicKey.slice(0, 8).toString('hex');
            console.log('  Key pair', i + 1, 'public key prefix:', firstBytes);
        }

        // Step 9: Demonstrate use with encryption
        console.log('');
        console.log('Step 9: Using generated keys for encryption...');
        const message = 'This message will be encrypted with the generated key pair.';
        const textBuffer = Buffer.from(message, 'utf8');

        // Encrypt using the public key we exported
        const encryptor = new zupt.Encryptor(publicKey);
        const { ciphertext, encHeader } = encryptor.encryptMemory(textBuffer);

        console.log('  Encrypted message size:', ciphertext.length, 'bytes');

        // Decrypt using the loaded private key
        const decryptor = new zupt.Decryptor(loadedKeyPair.secretKey);
        const decrypted = decryptor.decryptMemory(ciphertext, encHeader);
        const decryptedMessage = decrypted.toString('utf8');

        if (decryptedMessage === message) {
            console.log('  SUCCESS: Encryption/decryption with loaded keys works!');
        } else {
            console.log('  ERROR: Decrypted message does not match!');
        }

        console.log('');
        console.log('All key generation examples completed successfully!');

    } catch (err) {
        console.error('Error:', err.message);
        if (err.stack) {
            console.error(err.stack);
        }
        process.exit(1);
    } finally {
        // Cleanup
        try {
            fs.unlinkSync(path.join(testDir, 'keys.pem'));
            fs.unlinkSync(path.join(testDir, 'keys.pub.pem'));
            fs.rmdirSync(testDir);
        } catch (e) {
            // Ignore cleanup errors
        }
    }
}

main();

File: examples_rust/examples/example_keygen.rs
use libzupt_rust;
use std::fs;

fn main() {
    println!("{}", "=".repeat(60));
    println!("libzupt - Key Generation and Management Example (Rust)");
    println!("{}", "=".repeat(60));
    println!();

    let tmpdir = "rust_zupt_test_keys";
    fs::create_dir_all(tmpdir).expect("Failed to create temp dir");
    let priv_key_file = format!("{}/private.key", tmpdir);
    let pub_key_file = format!("{}/public.key", tmpdir);

    println!("1. Generating key pair...");
    let keypair = libzupt_rust::generate_keypair().expect("Failed to generate key pair");
    println!("   Public key: {} bytes", keypair.public_key.len());
    println!("   Private key: {} bytes", keypair.secret_key.len());
    println!();

    println!("2. Saving key pair...");
    libzupt_rust::save_keypair(&keypair, &priv_key_file).expect("Failed to save key pair");
    println!("   Saved to: {}", priv_key_file);
    println!();

    println!("3. Exporting public key...");
    let mut exported_pub = vec![0u8; libzupt_rust::HYBRID_PUB_KEY_SIZE];
    libzupt_rust::export_public_key(&keypair.secret_key, &mut exported_pub)
        .expect("Failed to export public key");
    libzupt_rust::write_file(&pub_key_file, &exported_pub).expect("Failed to write public key");
    println!("   Saved to: {}", pub_key_file);
    println!();

    println!("4. Loading key pair...");
    let loaded_kp = libzupt_rust::load_keypair(&priv_key_file).expect("Failed to load key pair");
    println!("   Loaded public key: {} bytes", loaded_kp.public_key.len());
    println!("   Loaded private key: {} bytes", loaded_kp.secret_key.len());
    assert_eq!(loaded_kp.public_key, keypair.public_key);
    assert_eq!(loaded_kp.secret_key, keypair.secret_key);
    println!("   Keys match!");
    println!();

    println!("5. Loading public key only...");
    let loaded_pub = libzupt_rust::load_public_key(&pub_key_file).expect("Failed to load public key");
    println!("   Loaded public key: {} bytes", loaded_pub.len());
    assert_eq!(loaded_pub, keypair.public_key);
    println!("   Public key matches!");
    println!();

    println!("6. Key sizes (bytes):");
    println!("   ML-KEM public key: {}", libzupt_rust::MLKEM_PUBLICKEYBYTES);
    println!("   X25519 public key: {}", libzupt_rust::X25519_KEYBYTES);
    println!("   Hybrid public key: {}", libzupt_rust::HYBRID_PUB_KEY_SIZE);
    println!("   Hybrid private key: {}", libzupt_rust::HYBRID_PRIV_KEY_SIZE);
    println!("   Encryption header: {}", libzupt_rust::HYBRID_ENC_HEADER_SIZE);
    println!();

    fs::remove_dir_all(tmpdir).ok();

    println!("{}", "=".repeat(60));
    println!("Key management example passed!");
    println!("{}", "=".repeat(60));
}
